### PR TITLE
Added functions for when a user joins any channel, parts the current channel, and joins/parts the server.

### PR DESCRIPTION
**Pull request by narenniranjan**
dude functionality lmao

### CODE REVIEW

File: mumble/client.py [hunk 1]
```diff
@@ -112,6 +112,22 @@ def request_blobs(self, texture_for_users=None, comment_for_users=None,
             channel_descriptions=[channel.id
                                   for channel in description_for_channels])
 
+    def user_join_channel(self, user, channel):
```
Comment (rfw) [inline mumble/client.py]: can you call this user_joined_channel? same for everything below

File: mumble/entities.py [hunk 2]
```diff
@@ -115,8 +115,10 @@ def __init__(self, client, session):
         super().__init__()
         self.client = client
         self.session = session
+        self.channel_id = 0
 
     def get_channel(self):
+        print(self.channel_id)
```
Comment (rfw) [inline mumble/entities.py]: delete

File: mumble/client.py [hunk 3]
```diff
@@ -112,6 +112,22 @@ def request_blobs(self, texture_for_users=None, comment_for_users=None,
...
+        # Override me!
+        pass
+
+    def user_part_channel(self, user, channel):
+        # Override me!
+        pass
+
+    def user_join_server(self, user):
```
Comment (rfw) [inline mumble/client.py]: user_connected/user_disconnected?

File: mumble/client.py [hunk 4]
```diff
@@ -146,9 +162,22 @@ def control_channel_remove_received(self, channel_id):
...
+            new_user = True
+        else:
+            new_user = False
+            old_chan = self.users[state.session].channel_id
         self._update_user(state)
+        user = self.users[state.session]
+        if new_user:
+            self.user_join_server(user)
+        else:
+            self.user_join_channel(user, user.channel_id)
```
Comment (rfw) [inline mumble/client.py]: can you pass the channel object in here? e.g. self.channels[user.channel_id]

File: mumble/client.py [hunk 5]
```diff
@@ -146,9 +162,22 @@ def control_channel_remove_received(self, channel_id):
...
+            if old_chan == self.me.get_channel().id and old_chan != user.channel_id:
```
Comment (rfw) [inline mumble/client.py]: don't check against your own channel here; the client code should be responsible for checking which channel

Comment (narenniranjan) [inline mumble/client.py]: erm, what do you mean? Do you mean I should report all parts from all channels, and then do the checking in user_part_channel()?

File: mumble/client.py [hunk 6]
```diff
@@ -146,9 +158,23 @@ def control_channel_remove_received(self, channel_id):
...
+            new_user = True
+        else:
+            new_user = False
+            old_chan = self.users[state.session].channel_id
         self._update_user(state)
+        user = self.users[state.session]
+        if new_user:
+            self.user_connect(user)
+            self.user_move_channel(user, None, self.channels[user.channel_id])
+        elif self.channels[old_chan] != self.channels[user.channel_id]:
```
Comment (rfw) [inline mumble/client.py]: use `is not` here

File: mumble/client.py [hunk 7]
```diff
@@ -112,6 +112,18 @@ def request_blobs(self, texture_for_users=None, comment_for_users=None,
             channel_descriptions=[channel.id
                                   for channel in description_for_channels])
 
+    def user_move_channel(self, user, source, dest):
```
Comment (rfw) [inline mumble/client.py]: user_moved

File: mumble/client.py [hunk 8]
```diff
@@ -112,6 +112,18 @@ def request_blobs(self, texture_for_users=None, comment_for_users=None,
...
+        # Override me!
+        pass
+
+    def user_connect(self, user):
```
Comment (rfw) [inline mumble/client.py]: user_connected

File: mumble/client.py [hunk 9]
```diff
@@ -112,6 +112,18 @@ def request_blobs(self, texture_for_users=None, comment_for_users=None,
...
+        # Override me!
+        pass
+
+    def user_disconnect(self, user):
```
Comment (rfw) [inline mumble/client.py]: user_disconnected

**Pull request merged**